### ISSUE TITLE
watchpoint support for (global) variables

### ISSUE THREAD

**Issue by namhyung** [namhyung/uftrace#661]
This supports watchpoint to track global variables (and raw addresses).  Only pritimitive integer types are supported for now.  Larger types or struct will be ignored.

The syntax is like following (to watch "myvar" variable):

```
$ uftrace -W var:myvar  myprog
```

Optional type size can be added like arguments.

I used the below example code for testing:

```c
$ cat watch-test.c 
#include <stdio.h>

int a = -1;
int mybar = 7;
int b = -1;

void bar(void);

void foo(void) {
  myvar = 2;
  bar();
  myvar = 4;
}

void bar(void) {
  a = 0;
}

int main(void) {
  a = 1;
  foo();
  b = 5;
  return 0;
}
```

And tested with following command:

```
$ uftrace -W var:myvar  watch-test
# DURATION     TID     FUNCTION
   3.194 us [ 17079] | __monstartup();
   0.443 us [ 17079] | __cxa_atexit();
            [ 17079] | main() {
            [ 17079] |   foo() {
            [ 17079] |     /* watch:var (myvar = 2) */
   2.983 us [ 17079] |     bar();
            [ 17079] |     /* watch:var (myvar = 4) */
   4.626 us [ 17079] |   } /* foo */
   5.913 us [ 17079] | } /* main */
```

One more test to track another variable

```
$ uftrace -W var:a  watch-test
# DURATION     TID     FUNCTION
   2.433 us [ 17660] | __monstartup();
   0.372 us [ 17660] | __cxa_atexit();
            [ 17160] | main() {
            [ 17160] |   /* watch:var (a = 1) */
            [ 17160] |   foo() {
            [ 17160] |     bar() {
            [ 17160] |       /* watch:var (a = 0) */
   0.270 us [ 17160] |     } /* bar */
   4.437 us [ 17160] |   } /* foo */
   5.004 us [ 17160] | } /* main */
```

It's possible to add a watch point by using address (for some reason)

```
$ uftrace --no-randomize-addr -W addr:55555555802c/i32  watch-test
# DURATION     TID     FUNCTION
   8.610 us [ 17197] | __monstartup();
   0.474 us [ 17197] | __cxa_atexit();
            [ 17197] | main() {
            [ 17197] |   foo() {
            [ 17197] |     /* watch:addr (addr:55555555802c = 2) */
   3.125 us [ 17197] |     bar();
            [ 17197] |     /* watch:addr (addr:55555555802c = 4) */
   4.695 us [ 17197] |   } /* foo */
   5.909 us [ 17197] | } /* main */
```

The code is available at [review/watch-variable-v1](https://github.com/namhyung/uftrace/compare/review/watch-variable-v1).

**Comment by honggyukim** [comment]
Looks great! But what about showing both old and new values together?

**Comment by namhyung** [comment]
Do you really need it?  I thought it might be annoying if value can be long (like strings - although not supported now).

**Comment by honggyukim** [comment]
I was thinking `gdb`'s watch feature that prints both old and new values.

But now, I think that the previous value might be wrong because the value might be changed many more times inside because we just check at the function boundaries only. So I'm okay with the current output.

**Comment by honggyukim** [comment]
@namhyung What about supporting a way to register all the global variables in the watchpoint list?
It may show some interesting tracing results.
I know that you don't want to add more options but `--watch-all` could be a good candidate for this.

**Comment by namhyung** [comment]
I think it also supports pattern matching (at least eventually)

**Comment by honggyukim** [comment]
> I think it also supports pattern matching (at least eventually)

So I could use `-W.` similar to other filters. It's not supported yet but hope to have it as well. Thanks.

**Comment by namhyung** [comment]
Updated [review/watch-variable-v2](../compare/review/watch-variable-v2).

**Comment by namhyung** [comment]
Uploaded [review/watch-variable-v3](../compare/review/watch-variable-v3).

Changes:
* show initial value
* add a test case

```
$ uftrace -W var:mydata tests/t-watch-global
# DURATION     TID     FUNCTION
            [243156] | __monstartup() {
            [243156] |   /* watch:var (mydata=0) */
   2.803 us [243156] | } /* __monstartup */
   0.621 us [243156] | __cxa_atexit();
            [243156] | main() {
            [243156] |   foo() {
            [243156] |     /* watch:var (mydata=1) */
   0.117 us [243156] |     bar();
            [243156] |     /* watch:var (mydata=2) */
   0.643 us [243156] |   } /* foo */
   0.938 us [243156] | } /* main */
```

**Comment by namhyung** [comment]
Updated: [review/watch-variable-v4](../compare/review/watch-variable-v4).

Changes:
 * drop 'address' watchpoint both for simplicity and safety.

**Comment by namhyung** [comment]
Merged 3ad5f45b.

**Comment by honggyukim** [comment]
Sorry, I couldn't test it enough but thanks for the update.
